README.md
# antd-spin

[![NPM version](https://img.shields.io/npm/v/antd-spin)](https://www.npmjs.com/package/antd-spin)
![Downloads](https://img.shields.io/npm/dw/antd-spin)
[![License](https://img.shields.io/npm/l/antd-spin)](./LICENSE)

> Show animation while loading data.

### 目的

如果你使用 Vue 开发项目，那么你一定用过或听过大名鼎鼎的 [Element-UI](https://element.eleme.cn/#/zh-CN/component/installation)，在 [Element-UI](https://element.eleme.cn/#/zh-CN/component/installation) 众多好用的组件中，有一个组件叫 Loading 组件，这个组件使用起来特别的灵活，支持：

- 指令方式和服务方式（服务方式还带单例模式）
- 灵活的配置项，特别是 target 可以随意指定渲染节点。

可惜的是，强大无比的 Antd ，它的 [Spin](https://ant.design/components/spin-cn/) 组件竟然就只支持指令方式，而且配置选项还无法支持我们指定 DOM 进行渲染，尤其是在项目中使用非常的不方便。

所以就用 Antd UI 框架的 [Spin](https://ant.design/components/spin-cn/) 组件和 [Icon](https://ant.design/components/icon-cn/) 组件来实现 [Element-UI](https://element.eleme.cn/#/zh-CN/component/installation) 的 [Loading](https://element.eleme.cn/#/zh-CN/component/loading) 组件的**服务方式** 功能。

### 快速开始

1. 使用 npm 安装依赖

```bash
$ npm install antd --save-dev
```

2. 使用

```js
// 1. 引入，支持两种方式的引入
import antdSpin from "antd-spin";
// import { antdSpin } from "antd-spin";

// 2.创建一个实例，默认创建一个全局的 loading
/* 
    创建一个 loading 实例，未关闭这个实例之前，只能创建一个，因为它单例的。
    注意⚠️：service 的参数都是可选的
*/
const loadingInstance = antdSpin.service();
const loadingInstance1 = antdSpin.service();
const loadingInstance2 = antdSpin.service();
/* 只会有一个 loading 生效 */
console.log(loadingInstance1 === loadingInstance2); // true

// 3. 关闭
/* 五秒之后自动关闭 */
setTimeout(() => {
    loadingInstance.close();
}, 5000);

// 4. 浏览器刷新项目，观察
```

![singleton-mode](./img/singleton-mode.gif)

### 配置

基本配置主要包含一些公用的常用配置，如下：

| Property | Description | Type | Default Value|
| :---: | :---: | :---: | :---: |
| 参数 | 说明 | 类型 | 默认值 |
| text | 显示在加载图标下方的加载文案 | string | - |
| size | 组件大小，可选值为 `small` `default` `large` | string | `default` |
| background | 遮罩背景色 | string | - |
| customClass | Loading 组件的自定义类名 | string | - |
| lock | 出现 loading 的时候，禁止屏幕滚动 | boolean | false |
| log | 是否启用 `console` 打印日志 | boolean | false |
| target | Loading 需要覆盖的 DOM 节点。可传入一个 DOM 对象或字符串；若传入字符串，则会将其作为参数传入 document.querySelector 以获取到对应 DOM 节点 | object/string | document.body |

> 补充：target 字段支持 DOM 对象，DOM 对象支持两种方式使用。
- JS 原生 DOM 对象，例如通过 `document.getElementById("app")` 捕获的 DOM。
- React API 创建的 DOM 对象，例如通过 useRef/createRef 创建的 DOM 对象。记住给 antd-spin 传参的时候是 `ref.current`，而不是支持传入 `ref`。

加载图标的方法，主要有三种方式，注意⚠️，**三种加载方式只能任选其一，不能同时共存，否则会抛出错误。**：

1. 使用 Antd 自带的图标

| Property | Description | Type | Default Value|
| :---: | :---: | :---: | :---: |
| 参数 | 说明 | 类型 | 默认值 |
| indicator | 自定义加载图标 | string | - |

例如：要想加载 Antd 的 `<LoadingOutlined />` 这个图标组件，只需要这样配置：

```js
{
    ...options,
    indicator: "LoadingOutlined"
}
```
2. 自定义图标

如果项目有自己的图标，首先要保证这张图片是 SVG 格式的。看例子：

```js
//  1. 引入 SVG 图片
const HeartSvg = (props) => (
    <svg { ...props } width="1em" height="1em" fill="currentColor" viewBox="0 0 1024 1024">
        <path d="M923 283.6c-13.4-31.1-32.6-58.9-56.9-82.8-24.3-23.8-52.5-42.4-84-55.5-32.5-13.5-66.9-20.3-102.4-20.3-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5-24.4 23.9-43.5 51.7-56.9 82.8-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3 0.1-35.3-7-69.6-20.9-101.9z" />
    </svg>
);
/* 
    这里千万注意⚠️ ，自定义SVG图标函数，必须有 props 属性，才能继承 Antd 的 spin（动画） rotate（旋转）等属性。
    没有 props 的话，直接开启 Antd 的 spin/rotate 是没有效果的。
    坑爹的 Antd 官网竟然没有说明
    Icon spin prop not working when custom ：https://github.com/ant-design/ant-design-icons/issues/270
*/

// 2. 通过 component 字段配置使用
{
    ...options,
    component: HeartSvg
}
```

3. 在线图标（不推荐）

> 友情提示：因为这种图标是外链阿里妈妈的网站，所以图标加载视网络速度可能会出现不同时间的延迟，所以不推荐使用，有此功能纯碎是为了支持 Antd 的 Icon 组件的所有使用方式。

如果你项目的图标是通过 [iconfont](https://www.iconfont.cn/) 来管理的，可直接在项目中外链引入网站上的图标：

```js
{
    ...options,
    IconFont: {
        type: "icon-tuichu",
        scriptUrl: "//at.alicdn.com/t/font_8d5l8fzk5b87iudi.js"
    }
}

// 要想引入多个资源， scriptUrl 当成数组使用
{
    ...options,
    IconFont: {
        type: "icon-tuichu",
        scriptUrl: [
            "//at.alicdn.com/t/font_1788044_0dwu4guekcwr.js", // icon-javascript, icon-java, icon-shoppingcart (overrided)
            "//at.alicdn.com/t/font_1788592_a5xf2bdic3u.js", // icon-shoppingcart, icon-python
        ]
    }
}

```

上面三种引入图标的方式都可以通过 loadingConfig 字段来添加一些自定义的动画，样式等。loadingConfig 字段所有的默认配置如下：

```js
{
    ...options,
    // 加载图标的一些配置
    loadingConfig: {
        // 自定义加载图标类名
        spinner: "",
        /* 图标旋转角度（IE9 无效） */
        rotate: 0,
        /* 是否有旋转动画 */
        spin: false,
        /* style 设置图标的样式，例如 fontSize 和 color */
        style: {},
        /* 仅适用双色图标。设置双色图标的主要颜色	string (十六进制颜色) */
        twoToneColor: "#eb2f96"
    }
}
```

> 当前时间 Tuesday, November 24, 2020 15:59:25 

CHANGELOG.md
# CHANGELOG

## v1.0.6 (2020-11-27 11:34:20)

fix: HMR error and IconFont add loadingConfig property

## v1.0.5 (2020-11-27 01:43:05)

fix: option default target value "" => null

## v1.0.4 (2020-11-27 01:20:15)

chore: add can be cancelled log

## v1.0.3 (2020-11-27 01:01:55)

fix: options target param bug

## v1.0.2 (2020-11-26 22:18:40)

fix: change DOM judgment mode

## v1.0.1 (2020-11-24 16:41:35)

public release

## v1.0.0 (2020-11-12 01:37:35)

Initial public release

src/components/antdSpin.js
import { Spin } from "antd";
import React from "react";
import ReactDOM from "react-dom";
import "./index.css";
import Icon, { createFromIconfontCN } from "@ant-design/icons";

class antdSpin {
  options = {};
  dom = null;
  targetDOM = null;
  // 当前是否正在请求
  requestFlag = false;
  config = {
    /*
            可传入一个 DOM 对象或字符串；
            若传入字符串，则会将其作为参数传入 document.querySelector以获取到对应 DOM 节点
        */
    target: null,
    lock: false,
    // 显示在加载图标下方的加载文案
    text: "",
    background: "transparent",
    // 组件大小，可选值为 small default large
    size: "large",
    // 自定义组件的类名
    customClass: "",
    // 加载图标的一些配置
    loadingConfig: {
      // 自定义加载图标类名
      spinner: "",
      /* 图标旋转角度（IE9 无效） */
      rotate: 0,
      /* 是否有旋转动画 */
      spin: false,
      /* style 设置图标的样式，例如 fontSize 和 color */
      style: {},
      /* 仅适用双色图标。设置双色图标的主要颜色	string (十六进制颜色) */
      twoToneColor: "#eb2f96",
    },
    indicator: "",
    /* 使用 SVG 自定义的图标 */
    component: null,
    IconFont: {},
    // 取消 console.log
    log: false,
  };

  service(options = {}) {
    /* 先判断 target 字段是否合法 */
    if (options?.hasOwnProperty("target")) {
      const targetValue = options.target;
      if (
        typeof targetValue === "string" ||
        targetValue instanceof HTMLElement ||
        targetValue === null ||
        targetValue?.hasOwnProperty("current")
      ) {
        this.options.log && console.log("option.target is effective!");
      } else {
        throw new Error("option target error,please check!");
      }
    }
    /* options 未自定义的字段，使用默认值 */
    const config = this.config;
    for (let k in config) {
      if (!options[k]) {
        options[k] = config[k];
      }

      if (k === "loadingConfig" && options[k]) {
        for (let j in config[k]) {
          if (!options[k][j]) {
            options[k][j] = config[k][j];
          }
        }
      }
    }

    /* 备份，主要方便 close 使用 */
    this.options = options;
    this.targetDOM = options.target;

    /* 单例模式，保证未关闭前只打开一个 loading */
    if (!this.requestFlag) {
      this.requestFlag = true;
      // 创建 Spin/loading
      this.dom = document.createElement("div");

      /* 添加蒙版 */
      this.dom.classList.add("el-loading-common-mask");
      /* 如果添加到全局，定位采用 fixed，否则采用 abso */
      if (this.targetDOM === null) {
        this.dom.classList.add("el-loading-fixed-mask");
      } else {
        this.dom.classList.add("el-loading-absolute-mask");
      }

      /* 蒙版背景色 */
      this.dom.setAttribute("style", `background-color: ${options.background};`);

      /* 是否添加自定义类名 */
      if (options.customClass.length) {
        this.dom.classList.add(options.customClass);
      }

      /* 是否禁止滚动 */
      if (options.lock) {
        document.documentElement.classList.add("antd-loading-html--hidden");
      }

      /*
                难点
            */
      this.customSpinIcons().then((AntDesignIcons) => {
        ReactDOM.render(<Spin tip={options.text} size={options.size} indicator={AntDesignIcons} />, this.dom);
      });

      /*
                1. target 为字符串，使用 querySelector 获取 DOM
                2. target 是 DOM，JS 原生 DOM 或 ReactNode
            */
      if (typeof options.target === "string" && options.target.length) {
        this.targetDOM = document.querySelector(`#${options.target}`) || document.querySelector(`.${options.target}`);

        /* 判断得到是否是 DOM 节点，不是的话 直接报错 */
        if (this.targetDOM) {
          this.appendDOM2Target();
        } else {
          throw new Error("find error: No DOM found using querySelector API!");
        }
      } else if (this.targetDOM?.hasOwnProperty("current")) {
        throw new Error(
          "when u use ReactDOM's method to pass reference to target, please input 👉ref.current👈 instead of 👉ref👈"
        );
      } else if (this.targetDOM instanceof HTMLElement) {
        // 1. ReactDOM 2. JS 原生 DOM
        this.appendDOM2Target();
      } else {
        document.body.appendChild(this.dom);
      }
    }

    return this;
  }

  async customSpinIcons() {
    let loadingConfig = this.options.loadingConfig ?? {};
    /*
            图标分三种：
            1. antd 自带的基本图标 indicator 字段控制
            1. 用户自定义的图标 component 字段控制
            2. 在线 iconfont 图标 IconFont 字段控制
            设置时，只能三选一
        */
    if (this.options.IconFont.type || this.options.IconFont.scriptUrl) {
      if (this.options.component || this.options.indicator.length) {
        throw new Error(
          "Icon configuration field：You set the ⭐️IconFont⭐️ field has been effective, please delete 👉component and indicator👈！"
        );
      }
    }

    if (this.options.component) {
      if (this.options.IconFont.type || this.options.IconFont.scriptUrl || this.options.indicator.length) {
        throw new Error(
          "Icon configuration field：You set the ⭐️component⭐️ field has been effective, please delete 👉IconFont and indicator👈！"
        );
      }
    }

    if (this.options.indicator.length) {
      if (this.options.IconFont.type || this.options.IconFont.scriptUrl || this.options.component) {
        throw new Error(
          "Icon configuration field：You set the ⭐️indicator⭐️ field has been effective, please delete 👈component and 👉IconFont👈！"
        );
      }
    }

    if (this.options.IconFont.type) {
      const { type, scriptUrl } = this.options.IconFont;
      const IconFont = createFromIconfontCN({ scriptUrl });
      /* IconFont 图标有 twoToneColor 会报错，所以需要去除它 */
      delete loadingConfig.twoToneColor;
      return <IconFont type={type} {...loadingConfig} />;
    }

    if (this.options.component) {
      /* 自定义图标有 twoToneColor 会报错，所以需要去除它 */
      delete loadingConfig.twoToneColor;

      const CustomIcon = (props) => <Icon component={this.options.component} {...props} />;
      return <CustomIcon {...loadingConfig} />;
    }

    if (this.options.indicator.length) {
      const modules = await import("@ant-design/icons");
      const AntDesignIcons = modules[this.options.indicator];
      if (!AntDesignIcons) {
        throw new Error("sorry the module you need was not found in @ant-design/icons4!");
      }

      return <AntDesignIcons {...loadingConfig} />;
    } else {
      return "";
    }
  }

  appendDOM2Target() {
    /* 定位 子绝父相 */
    // static | relative | absolute | sticky | fixed
    const position = globalThis.getComputedStyle(this.targetDOM, null).getPropertyValue("position");
    if (position === "inherit" || position === "static") {
      /* 还需要移除 */
      this.targetDOM.classList.add("antd-targetDOM-position");
    }
    this.targetDOM.appendChild(this.dom);
  }

  domRemoveChild(containerDOM, targetDOM) {
    if (containerDOM?.contains(targetDOM)) {
      containerDOM.removeChild(targetDOM);
      this.options.log && console.log("ChildDOM remove success!");
    } else {
      this.options.log && console.log("ChildDOM not append containerDOM!");
    }
  }

  close() {
    /* 移除已经上树的 React 组件 Spin，并清除其绑定的事件和状态 */
    const unmountComponentAtNode = ReactDOM.unmountComponentAtNode(this.dom);
    if (unmountComponentAtNode) {
      this.options.log && console.info("component was unmounted!");
    } else {
      this.options.log && console.info("no component to unmount!");
    }

    // 隐藏 Spin/loading
    if (this.requestFlag) {
      this.requestFlag = false;
      /* 移除 DOM */
      if (this.targetDOM !== null) {
        this.domRemoveChild(this.targetDOM, this.dom);
        this.targetDOM.classList.remove("antd-targetDOM-position");
      } else {
        this.domRemoveChild(document.body, this.dom);
      }
      this.targetDOM = this.dom = null;
    }

    /* 禁止滚动时，需要移除 */
    if (this.options.lock) {
      this.options.lock = false;
      document.documentElement.classList.remove("antd-loading-html--hidden");
    }
  }
}

export default new antdSpin();


lib/index.js
import { Spin } from 'antd';
import React from 'react';
import ReactDOM from 'react-dom';
import Icon, { createFromIconfontCN } from '@ant-design/icons';

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
  try {
    var info = gen[key](arg);
    var value = info.value;
  } catch (error) {
    reject(error);
    return;
  }

  if (info.done) {
    resolve(value);
  } else {
    Promise.resolve(value).then(_next, _throw);
  }
}

function _asyncToGenerator(fn) {
  return function () {
    var self = this,
        args = arguments;
    return new Promise(function (resolve, reject) {
      var gen = fn.apply(self, args);

      function _next(value) {
        asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
      }

      function _throw(err) {
        asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
      }

      _next(undefined);
    });
  };
}

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".el-loading-common-mask {\n  top: 0;\n  left: 0;\n  right: 0;\n  bottom: 0;\n  background: transparent;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  z-index: 9999;\n  font-size: 20px;\n}\n\n.el-loading-absolute-mask {\n  position: absolute;\n}\n\n.el-loading-fixed-mask {\n  position: fixed;\n}\n\n.antd-loading-html--hidden {\n  overflow: hidden !important;\n}\n\n.antd-targetDOM-position {\n  position: relative;\n}\n";
styleInject(css_248z);

var antdSpin = /*#__PURE__*/function () {
  function antdSpin() {
    _classCallCheck(this, antdSpin);

    _defineProperty(this, "options", {});

    _defineProperty(this, "dom", null);

    _defineProperty(this, "targetDOM", null);

    _defineProperty(this, "requestFlag", false);

    _defineProperty(this, "config", {
      /*
              可传入一个 DOM 对象或字符串；
              若传入字符串，则会将其作为参数传入 document.querySelector以获取到对应 DOM 节点
          */
      target: null,
      lock: false,
      // 显示在加载图标下方的加载文案
      text: "",
      background: "transparent",
      // 组件大小，可选值为 small default large
      size: "large",
      // 自定义组件的类名
      customClass: "",
      // 加载图标的一些配置
      loadingConfig: {
        // 自定义加载图标类名
        spinner: "",

        /* 图标旋转角度（IE9 无效） */
        rotate: 0,

        /* 是否有旋转动画 */
        spin: false,

        /* style 设置图标的样式，例如 fontSize 和 color */
        style: {},

        /* 仅适用双色图标。设置双色图标的主要颜色	string (十六进制颜色) */
        twoToneColor: "#eb2f96"
      },
      indicator: "",

      /* 使用 SVG 自定义的图标 */
      component: null,
      IconFont: {},
      // 取消 console.log
      log: false
    });
  }

  _createClass(antdSpin, [{
    key: "service",
    value: function service() {
      var _this = this;

      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      /* 先判断 target 字段是否合法 */
      if (options !== null && options !== void 0 && options.hasOwnProperty("target")) {
        var targetValue = options.target;

        if (typeof targetValue === "string" || targetValue instanceof HTMLElement || targetValue === null || targetValue !== null && targetValue !== void 0 && targetValue.hasOwnProperty("current")) {
          this.options.log && console.log("option.target is effective!");
        } else {
          throw new Error("option target error,please check!");
        }
      }
      /* options 未自定义的字段，使用默认值 */


      var config = this.config;

      for (var k in config) {
        if (!options[k]) {
          options[k] = config[k];
        }

        if (k === "loadingConfig" && options[k]) {
          for (var j in config[k]) {
            if (!options[k][j]) {
              options[k][j] = config[k][j];
            }
          }
        }
      }
      /* 备份，主要方便 close 使用 */


      this.options = options;
      this.targetDOM = options.target;
      /* 单例模式，保证未关闭前只打开一个 loading */

      if (!this.requestFlag) {
        var _this$targetDOM;

        this.requestFlag = true; // 创建 Spin/loading

        this.dom = document.createElement("div");
        /* 添加蒙版 */

        this.dom.classList.add("el-loading-common-mask");
        /* 如果添加到全局，定位采用 fixed，否则采用 abso */

        if (this.targetDOM === null) {
          this.dom.classList.add("el-loading-fixed-mask");
        } else {
          this.dom.classList.add("el-loading-absolute-mask");
        }
        /* 蒙版背景色 */


        this.dom.setAttribute("style", "background-color: ".concat(options.background, ";"));
        /* 是否添加自定义类名 */

        if (options.customClass.length) {
          this.dom.classList.add(options.customClass);
        }
        /* 是否禁止滚动 */


        if (options.lock) {
          document.documentElement.classList.add("antd-loading-html--hidden");
        }
        /*
                  难点
              */


        this.customSpinIcons().then(function (AntDesignIcons) {
          ReactDOM.render( /*#__PURE__*/React.createElement(Spin, {
            tip: options.text,
            size: options.size,
            indicator: AntDesignIcons
          }), _this.dom);
        });
        /*
                  1. target 为字符串，使用 querySelector 获取 DOM
                  2. target 是 DOM，JS 原生 DOM 或 ReactNode
              */

        if (typeof options.target === "string" && options.target.length) {
          this.targetDOM = document.querySelector("#".concat(options.target)) || document.querySelector(".".concat(options.target));
          /* 判断得到是否是 DOM 节点，不是的话 直接报错 */

          if (this.targetDOM) {
            this.appendDOM2Target();
          } else {
            throw new Error("find error: No DOM found using querySelector API!");
          }
        } else if ((_this$targetDOM = this.targetDOM) !== null && _this$targetDOM !== void 0 && _this$targetDOM.hasOwnProperty("current")) {
          throw new Error("when u use ReactDOM's method to pass reference to target, please input 👉ref.current👈 instead of 👉ref👈");
        } else if (this.targetDOM instanceof HTMLElement) {
          // 1. ReactDOM 2. JS 原生 DOM
          this.appendDOM2Target();
        } else {
          document.body.appendChild(this.dom);
        }
      }

      return this;
    }
  }, {
    key: "customSpinIcons",
    value: function () {
      var _customSpinIcons = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var _this$options$loading,
            _this2 = this;

        var loadingConfig, _this$options$IconFon, type, scriptUrl, IconFont, CustomIcon, modules, AntDesignIcons;

        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                loadingConfig = (_this$options$loading = this.options.loadingConfig) !== null && _this$options$loading !== void 0 ? _this$options$loading : {};
                /*
                        图标分三种：
                        1. antd 自带的基本图标 indicator 字段控制
                        1. 用户自定义的图标 component 字段控制
                        2. 在线 iconfont 图标 IconFont 字段控制
                        设置时，只能三选一
                    */

                if (!(this.options.IconFont.type || this.options.IconFont.scriptUrl)) {
                  _context.next = 4;
                  break;
                }

                if (!(this.options.component || this.options.indicator.length)) {
                  _context.next = 4;
                  break;
                }

                throw new Error("Icon configuration field：You set the ⭐️IconFont⭐️ field has been effective, please delete 👉component and indicator👈！");

              case 4:
                if (!this.options.component) {
                  _context.next = 7;
                  break;
                }

                if (!(this.options.IconFont.type || this.options.IconFont.scriptUrl || this.options.indicator.length)) {
                  _context.next = 7;
                  break;
                }

                throw new Error("Icon configuration field：You set the ⭐️component⭐️ field has been effective, please delete 👉IconFont and indicator👈！");

              case 7:
                if (!this.options.indicator.length) {
                  _context.next = 10;
                  break;
                }

                if (!(this.options.IconFont.type || this.options.IconFont.scriptUrl || this.options.component)) {
                  _context.next = 10;
                  break;
                }

                throw new Error("Icon configuration field：You set the ⭐️indicator⭐️ field has been effective, please delete 👈component and 👉IconFont👈！");

              case 10:
                if (!this.options.IconFont.type) {
                  _context.next = 15;
                  break;
                }

                _this$options$IconFon = this.options.IconFont, type = _this$options$IconFon.type, scriptUrl = _this$options$IconFon.scriptUrl;
                IconFont = createFromIconfontCN({
                  scriptUrl: scriptUrl
                });
                /* IconFont 图标有 twoToneColor 会报错，所以需要去除它 */

                delete loadingConfig.twoToneColor;
                return _context.abrupt("return", /*#__PURE__*/React.createElement(IconFont, _extends({
                  type: type
                }, loadingConfig)));

              case 15:
                if (!this.options.component) {
                  _context.next = 19;
                  break;
                }

                /* 自定义图标有 twoToneColor 会报错，所以需要去除它 */
                delete loadingConfig.twoToneColor;

                CustomIcon = function CustomIcon(props) {
                  return /*#__PURE__*/React.createElement(Icon, _extends({
                    component: _this2.options.component
                  }, props));
                };

                return _context.abrupt("return", /*#__PURE__*/React.createElement(CustomIcon, loadingConfig));

              case 19:
                if (!this.options.indicator.length) {
                  _context.next = 29;
                  break;
                }

                _context.next = 22;
                return import('@ant-design/icons');

              case 22:
                modules = _context.sent;
                AntDesignIcons = modules[this.options.indicator];

                if (AntDesignIcons) {
                  _context.next = 26;
                  break;
                }

                throw new Error("sorry the module you need was not found in @ant-design/icons4!");

              case 26:
                return _context.abrupt("return", /*#__PURE__*/React.createElement(AntDesignIcons, loadingConfig));

              case 29:
                return _context.abrupt("return", "");

              case 30:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function customSpinIcons() {
        return _customSpinIcons.apply(this, arguments);
      }

      return customSpinIcons;
    }()
  }, {
    key: "appendDOM2Target",
    value: function appendDOM2Target() {
      /* 定位 子绝父相 */
      // static | relative | absolute | sticky | fixed
      var position = globalThis.getComputedStyle(this.targetDOM, null).getPropertyValue("position");

      if (position === "inherit" || position === "static") {
        /* 还需要移除 */
        this.targetDOM.classList.add("antd-targetDOM-position");
      }

      this.targetDOM.appendChild(this.dom);
    }
  }, {
    key: "domRemoveChild",
    value: function domRemoveChild(containerDOM, targetDOM) {
      if (containerDOM !== null && containerDOM !== void 0 && containerDOM.contains(targetDOM)) {
        containerDOM.removeChild(targetDOM);
        this.options.log && console.log("ChildDOM remove success!");
      } else {
        this.options.log && console.log("ChildDOM not append containerDOM!");
      }
    }
  }, {
    key: "close",
    value: function close() {
      /* 移除已经上树的 React 组件 Spin，并清除其绑定的事件和状态 */
      var unmountComponentAtNode = ReactDOM.unmountComponentAtNode(this.dom);

      if (unmountComponentAtNode) {
        this.options.log && console.info("component was unmounted!");
      } else {
        this.options.log && console.info("no component to unmount!");
      } // 隐藏 Spin/loading


      if (this.requestFlag) {
        this.requestFlag = false;
        /* 移除 DOM */

        if (this.targetDOM !== null) {
          this.domRemoveChild(this.targetDOM, this.dom);
          this.targetDOM.classList.remove("antd-targetDOM-position");
        } else {
          this.domRemoveChild(document.body, this.dom);
        }

        this.targetDOM = this.dom = null;
      }
      /* 禁止滚动时，需要移除 */


      if (this.options.lock) {
        this.options.lock = false;
        document.documentElement.classList.remove("antd-loading-html--hidden");
      }
    }
  }]);

  return antdSpin;
}();

var antdSpin$1 = new antdSpin();

export default antdSpin$1;
export { antdSpin$1 as antdSpin };
